File: lib/wType/bytearray.cpp
#include "bytearray.h"
#include <arpa/inet.h>

W::ByteArray::ByteArray(size_type size):
    data(new Container(size)),
    shiftBegin(0),
    shiftEnd(0),
    referenceMode(false)
{
    
}

W::ByteArray::ByteArray(const W::ByteArray& original):
    data(new Container(*(original.data))),
    shiftBegin(original.shiftBegin),
    shiftEnd(original.shiftEnd),
    referenceMode(original.referenceMode)
{

}

W::ByteArray::~ByteArray()
{
    delete data;
}

W::ByteArray& W::ByteArray::operator=(const W::ByteArray& original)
{
    if (&original != this)
    {
        delete data;
        data = new Container(*(original.data));
        shiftBegin = original.shiftBegin;
        shiftEnd = original.shiftEnd;
        referenceMode = original.referenceMode;
    }
    return *this;
}

W::ByteArray::size_type W::ByteArray::size() const
{
    return shiftEnd - shiftBegin;
}

void W::ByteArray::push8(uint8_t integer)
{
    if (referenceMode) {
        //TODO not sure
    }
    data->at(shiftEnd) = integer;
    ++shiftEnd;
}

void W::ByteArray::push16(uint16_t integer)
{
    uint16_t n16char = htons(integer);
    
    push8(((uint8_t*)&n16char)[0]);
    push8(((uint8_t*)&n16char)[1]);
}


void W::ByteArray::push32(uint32_t integer)
{
    uint32_t converted = htonl(integer);
    for (uint8_t i(0); i < 4; ++i)
    {
        push8(((uint8_t*)&converted)[i]);
    }
}

void W::ByteArray::push64(uint64_t integer)
{
    uint32_t h = integer >> 32;
    uint32_t l = integer & 0x00000000FFFFFFFF;
    
    uint32_t convertedh = htonl(h);
    for (uint8_t i(0); i < 4; ++i)
    {
        push8(((uint8_t*)&convertedh)[i]);
    }
    
    uint32_t convertedl = htonl(l);
    for (uint8_t i(0); i < 4; ++i)
    {
        push8(((uint8_t*)&convertedl)[i]);
    }
}


uint8_t W::ByteArray::pop8()
{
    uint8_t byte = data->at(shiftBegin);
    ++shiftBegin;
    return byte;
}

uint16_t W::ByteArray::pop16()
{
    uint8_t h = pop8();
    uint8_t l = pop8();
    
    uint8_t src[2] = {h, l};
    
    return ntohs(*((uint16_t*) src));
}


uint32_t W::ByteArray::pop32()
{
    uint8_t src[4];             //TODO optimize?
    
    for (uint8_t i(0); i < 4; ++i)
    {
        src[i] = pop8();
    }
    
    return ntohl(*((uint32_t*) src));
}

uint64_t W::ByteArray::pop64()
{
    uint8_t srch[4] = {0,0,0,0};
    uint8_t srcl[4] = {0,0,0,0};
    
    for (uint8_t i(0); i < 4; ++i)
    {
        srch[i] = pop8();
    }
    
    for (uint8_t i(0); i < 4; ++i)
    {
        srcl[i] = pop8();
    }
    
    uint64_t h = ntohl(*((uint32_t*) srch));
    uint32_t l = ntohl(*((uint32_t*) srcl));
    
    return (h << 32) | l;
}


char * W::ByteArray::getData()
{
    return data->data();        //TODO not actually sure about this, may be need to check first about shifts and refence?
}

W::ByteArray::size_type W::ByteArray::maxSize() const
{
    return data->size();
}

bool W::ByteArray::filled() const
{
    return shiftEnd == data->size();
}

W::ByteArray::size_type W::ByteArray::fill(const char* data, W::ByteArray::size_type size, W::ByteArray::size_type shift)
{
    size_type i = shift;
    while (i < size && !filled()) {
        push8(data[i]);
        ++i;
    }
    
    return i;
}

File: libjs/wType/bytearray.js
"use strict";
var Class = require("../utils/class");

var ByteArray = Class.inherit({
    "className": "ByteArray",
    "constructor": function(size) {
        Class.fn.constructor.call(this);
        
        this._referenceMode = false;
        this._data = new Uint8Array(size);
        this._shiftBegin = 0;
        this._shiftEnd = 0;
    },
    "_checkReference": function() {
        if (this._referenceMode) {
            var buffer = new ArrayBuffer(this._data.length - this._shiftBegin);
            var newData = new Uint8Array(buffer);
            newData.set(this._data, this._shiftBegin);
            this._data = newData;
            this._shiftBegin = 0;
            this._referenceMode = false;
        }
    },
    "fill": function(/*Uint8Array*/arr, /*Number*/size, /*[Number]*/shift) {
        this._checkReference();
        shift = shift || 0;
        
        if (this._shiftEnd === 0 && (this._data.length <= size - shift)) {
            this._referenceMode = true;
            this._data = arr.subarray(shift, this._data.length + shift);
            this._shiftEnd = this._data.length;
            shift += this._shiftEnd;
        } else {
            while (!this.filled() && shift < size) {
                this._data[this._shiftEnd] = arr[shift];
                ++shift;
                ++this._shiftEnd;
            }
        }
        
        return shift;
    },
    "filled": function() {
        return this._data.length === this._shiftEnd;
    },
    "size": function() {
        return this._shiftEnd - this._shiftBegin;
    },
    "maxSize": function() {
        return this._data.length;
    },
    "push8": function(int) {
        this._checkReference();
        
        this._data[this._shiftEnd] = int;
        ++this._shiftEnd;
    },
    "push16": function(int) {
        var h = (int >> 8) & 0xff;
        var l = int & 0xff;
        
        this.push8(h);
        this.push8(l);
    },
    "push32": function(int) {
        var hh = (int >> 24) & 0xff;
        var hl = (int >> 16) & 0xff;
        var lh = (int >> 8) & 0xff;
        var ll = int & 0xff;
        
        this.push8(hh);
        this.push8(hl);
        this.push8(lh);
        this.push8(ll);
    },
    "push64": function(int) {
        
    },
    "pop8": function(int) {
        var ret = this._data[this._shiftBegin];
        ++this._shiftBegin;
        return ret;
    },
    "pop16": function(int) {
        var ret = (this.pop8() << 8);
        ret = ret | this.pop8();
        
        return ret;
    },
    "pop32": function(int) {
        var ret = this.pop8() << 24;
        ret = ret | (this.pop8() << 16);
        ret = ret | (this.pop8() << 8);
        ret = ret | this.pop8();
        
        return ret;
    },
    "pop64": function(int) {
        
    },
    "data": function() {
        return this._data.subarray(this._shiftBegin, this._shiftEnd);
    }
});

module.exports = ByteArray;
